### PR TITLE
Close the HTTP pool when service is stopped. Take 2.

### LINKED ISSUE #86: stopService() leaves pooled connections open
With the default client, a persistent `HTTPConnectionPool` is created for connections to the ACME directory. This pool isn't tracked, however, and when `stopService()` is called on the txacme service a pooled connection remains waiting on the reactor. This breaks my integration tests.

### PR DESCRIPTION
**Pull request by glyph**
Scope
=====

This tries to fixes #86

When the service is stopped, it should automatically clean the resource and not left files opened or connections openend.

In Take1 the #154 regression was introduced.

Obviously there was lots of great stuff in code I had to revert out, so we should probably take a second crack at this.

This is the take 2 of https://github.com/twisted/txacme/pull/140

### CODE REVIEW

**Comment by glyph** [comment]
(I would suggest splitting this up into a bunch of smaller PRs, the PR template and contributing docs don't seem to have anything to do with the functional changes here.)

**Comment by adiroiban** [comment]
Thanks for the work on this.

I split it into a single other PR ... see here https://github.com/twisted/txacme/pull/157

I will look and try to fix the defect

**Comment by adiroiban** [comment]
@glyph this was a backward incompatible change.

As long as the fix for #150 is not merged and released I will leave this change unmerged,

I will add this change in my acme-v2 branch.... as that branch breaks the whole API.

**Comment by glyph** [comment]
> @glyph this was a backward incompatible change.

I'll have to address backwards incompatibility when I update `lancer` to deal with more recent versions; that's fine.  But that's not what I saw here: I didn't revert it because it was backwards incompatible; I reverted it because it was incompatible with *itself* (the still-included endpoint, which I'm still using in production despite the fact that the SNI challenge doesn't work, raised exceptions and refused to listen on the port).

> As long as the fix for #150 is not merged and released I will leave this change unmerged,

#150 is merged, I should do a release.

> I will add this change in my acme-v2 branch.... as that branch breaks the whole API.

Again, the incompatibilities are not a huge concern here; I'd much rather land this as a separate change if we can, just to make the review churn manageable.  (Although perhaps you are correct that we should do a release with #150 and without this, to provide a smoother upgrade path.)

Thank you for working on the acme-v2 branch though!  Is it already working for you, or just in progress?

**Comment by glyph** [comment]
> it was incompatible with _itself_

incompatible with itself *and also* the docs were never updated so they were still wrong :)

**Comment by adiroiban** [comment]
@glyph thanks for the info.

I tried to update the docstring for AutoTLSEndpoint.

Can you please take a look and let me know if you find any other errors?

Thanks!

------------

The acme v2 works for me ... and if you are just using the AcmeIssuingService it should be backward compatible with anyone.
The SAN is implemented using CSV names.

**Comment by adiroiban** [comment]
@glyph do you have time to check that the changes are ok now?

I want to have this merges as it can help with the future tests that I want to add for ACME v2 support.

Thanks!

**Comment by glyph** [comment]
@adiroiban I'll try to check it out this weekend; it would be great if you could address whatever's wrong on Travis before then, though!

**Comment by adiroiban** [comment]
I have retried the failings jobs... no is green (sort of :)

It looks like one failure is due to a flaky test while the other was a `pip install` error.

```
hypothesis.errors.Flaky: Hypothesis test_poll_timeout(self=<txacme.test.test_client.ClientTests.test_poll_timeout id=0x80de720>, name='daaa') produces unreliable results: Falsified on the first call but did not on a subsequent one
```

Thanks!

**Comment by glyph** [comment]
@adiroiban I know it's a bit of an artificial metric, but it seems a shame to sacrifice 100% code coverage.  Would you mind fixing up the coverage around `stop` and `tearDown` (in the latter case, just doing a coverage exclusion would be OK by me; it is test code after all) https://codecov.io/gh/twisted/txacme/compare/4288f5d876bb46b160db72406732c50875388eed...a474bd4c6bbc1561e4303a413aa2cd66d9b8fd99/diff#D3-814

File: src/txacme/testing.py [hunk 1]
```diff
@@ -19,6 +21,24 @@
 from txacme.util import clock_now, generate_private_key
 
 
+class TXACMETestCase(TestCase):
+    """
+    Common code for all tests for the txacme project.
```
Comment (glyph) [inline src/txacme/testing.py]: I definitely don't want all the tests in the whole project to have special nonstandard behavior :)

File: src/txacme/test/test_client.py [hunk 2]
```diff
@@ -344,6 +344,7 @@ def test_unexpected_update(self):
             self.assertThat(
                 client.register(reg2),
                 failed_with(IsInstance(errors.UnexpectedUpdate)))
+        succeeded(client.stop())
```
Comment (mithrandi) [inline src/txacme/test/test_client.py]: This constructs a matcher then does nothing with it; I think you want an `assertThat` call here.

File: src/txacme/testing.py [hunk 3]
```diff
@@ -19,6 +21,24 @@
 from txacme.util import clock_now, generate_private_key
 
 
+class TXACMETestCase(TestCase):
```
[the remark this answers is not in the document]
Comment (mithrandi) [reply src/txacme/testing.py]: In fact, I think you can just use `AsynchronousDeferredRunTest`; add `run_tests_with = AsynchronousDeferredRunTest` to the test cases needing this.

See https://testtools.readthedocs.io/en/latest/twisted-support.html#running-tests-in-the-reactor and https://github.com/testing-cabal/testtools/blob/209c1bf7618541f7b9a9e01798ef0588658d6ab1/testtools/twistedsupport/_runtest.py#L452

File: src/txacme/newsfragments/86.removal [hunk 4]
```diff
@@ -0,0 +1,4 @@
+INCOMPATIBLE CHANGE: txacme.client.JWSClient is now initialized with an
+twisted.web.client.Agent instead of treq.client.HTTPClient.
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply src/txacme/newsfragments/86.removal]: Yes... but Treq uses Twisted Agent for I/O... so just pass a stubbed Agent.

I think that stubbing a lower level API is a better practice.

At the same time, I think that Twisted should have a "standard" Agent stub which can be used by 3rd party software.
For my tests I am using InMemoryResponse and InMemoryPersistentAgent

File: src/txacme/test/test_client.py (hunk 2 above)
Comment (adiroiban) [reply src/txacme/test/test_client.py]: True. Thanks.

File: src/txacme/testing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adiroiban) [reply src/txacme/testing.py]: Happy to use a function... but the idea is that we want **all** tests to be executed with this check..and to make sure you don't forget to enable this check.

I agree that setUp / tearDown inheritance can be complicated, but at the same time I think that we should enforce that this check is called for all tests.

If we make the check optional, people might forget to enable it for the tests

File: src/txacme/newsfragments/86.removal (hunk 4 above)
[the remark this answers is not in the document]
Comment (glyph) [reply src/txacme/newsfragments/86.removal]: > At the same time, I think that Twisted should have a "standard" Agent stub which can be used by 3rd party software.

👍

File: src/txacme/testing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (glyph) [reply src/txacme/testing.py]: It's equally possible to forget to inherit from the test class though; like it's literally the same number of modified lines of code.  (one per `class` statement or one per `setUp` method).

However, there are other extension mechanisms we could use for this: the `test_suite` hook for example?

**Comment by glyph** [comment]
@adiroiban any chance you'll have some time to work on this soon?

**Comment by adiroiban** [comment]
I will not have time to work on this anytime soon. Feel free to continue the work.

I am using the code from this branch https://github.com/twisted/txacme/tree/151-acme-v2 and it works for me.

File: src/txacme/newsfragments/86.bugfix [hunk 5]
```diff
@@ -0,0 +1,4 @@
+INCOMPATIBLE CHANGE: txacme.service.AcmeIssuingService.stopFactory nows
+closes the persisted HTTP client connections.
+This is done to bring the in a state similar to the one before calling
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply src/txacme/newsfragments/86.bugfix:3]: I think that the last sentence can be removed.

**Comment by glyph** [comment]
I am going to just land this because we need to get v2 out for this to be useful at all, now that the LE endpoints have shut down new registrations and are starting brownouts.

**Comment by adiroiban** [comment]
I am sorry that I was not able to create a proper pull request.
But I think that share libraries should be as lightweight as possible.
For me eliot and hypothesis were a blocker as I am not familiar with those libraries and I didn't had time to learn how to use them.

I had the same issue that v2 should be delivered ASAP, so I went ahead and implemented txacme with v2 without eliot and with stdlib unit tests.

**Pull request merged**